xapian-core/api/valuerangeproc.cc

```diff
@@ -439,4 +439,82 @@ NumberRangeProcessor::operator()(const string& b, const string& e)
     return Xapian::Query(Xapian::Query::OP_INVALID);
 }
 
+static const string byte_units[4] = {
+    "B", "K", "M", "G"
+};
+
+// Return factor for byte unit
+// if string is a valid byte unit
+// else return -1
+static double
+check_byte_unit(const string &s) {
+	double factor = 1;
+	for (int i = 0; i < 4; ++i) {
+	    if (endswith(s, byte_units[i])) {
+		return factor;
+	    }
+	    factor *= 1024;
+	}
+
+	return -1;
+}
+
+Xapian::Query
+UnitRangeProcessor::operator()(const string& b, const string& e)
+{
+    // Parse the numbers to floating point.
+    double num_b, num_e;
+
+    if (!b.empty()) {
+	double factor_b = check_byte_unit(b);
+	if (factor_b == -1) {
+	    // Not a valid byte unit
+	    goto not_our_range;
+	}
+	string b_ = string(b, 0, b.size() - 1);
```

Comment:
There's no need to create a temporary string object here - `strtod()` will stop parsing when it reaches the units, so you can just work from `b` and adjust the end check to allow for the units.

Reply:
Updated.